### PR TITLE
Stripe add the ability to place hold on cards

### PR DESCRIPTION
**Pull request by joeauyeung**
## What does this PR do?

This PR allows the Stripe app to place holds on cards and allow users to charge a no-show fee to attendees.

image

image

image

image

Fixes # (issue)

**Environment**: Staging(main branch) / Production

## Type of change

- New feature (non-breaking change which adds functionality)

## How should this be tested?

- Ensure that you are forwarding Stripe webhooks locally
- Create an event type and add a price with the option to charge a no-show fee
- Create a booking and enter the Stripe testing card
    - Enter an invalid expiry date to render the error message on the payment page
- As the organizer go to the booking page
    - Charge the no-show fee to the attendee
    - There should be an email sent to the attendee
- Create a booking, but this time with a Stripe declined test card
- If you try to charge the no-show fee, an error message should appear

## Checklist

- I haven't added tests that prove my fix is effective or that my feature works
- I haven't checked if new and existing unit tests pass locally with my changes

### CODE REVIEW

File: packages/app-store/stripepayment/lib/PaymentService.ts [hunk 1]
```diff
@@ -62,6 +63,7 @@ export class PaymentService implements IAbstractPaymentService {
         currency: this.credentials.default_currency,
         payment_method_types: ["card"],
         application_fee_amount: paymentFee,
+        ...(paymentOptionEnum.parse(paymentOption) === "HOLD" && { capture_method: "manual" }),
```
Comment (joeauyeung) [inline packages/app-store/stripepayment/lib/PaymentService.ts:66]: By setting the `capture_method` of the payment intent to "manual" we can capture the funds in the future

File: packages/app-store/stripepayment/lib/PaymentService.ts [hunk 2]
```diff
@@ -92,6 +94,7 @@ export class PaymentService implements IAbstractPaymentService {
           fee: paymentFee,
           refunded: false,
           success: false,
+          paymentOption: paymentOptionEnum.parse(paymentOption) || "ON_BOOKING",
```
Comment (joeauyeung) [inline packages/app-store/stripepayment/lib/PaymentService.ts:97]: Most bookings by default should collect funds on booking

File: packages/app-store/_utils/getEventTypeAppData.ts [hunk 3]
```diff
@@ -28,6 +28,7 @@ export const getEventTypeAppData = <T extends EventTypeAppsList>(
       price: eventType.price,
       // Currency default is "usd" in DB.So, it would also be available always
       currency: eventType.currency,
+      paymentOption: "ON_BOOKING",
```
Comment (joeauyeung) [inline packages/app-store/_utils/getEventTypeAppData.ts:31]: I wasn't sure what to call the two payment scenarios. I wanted to reserve "payment method" if we ever want to accept other payments that were not cards. 

In this PR I want to introduce two options. `ON_BOOKING` is our normal flow where we collect the payment on booking.  The other option `HOLD` will be when we want to collect the payment information but not capture the funds yet.

File: packages/app-store/stripepayment/lib/constants.ts [hunk 4]
```diff
@@ -1,3 +1,14 @@
...
+export const paymentOptions = [
+  {
+    label: "on_booking_option",
+    value: "ON_BOOKING",
+  },
+  {
+    label: "hold_option",
+    value: "HOLD",
+  },
+];
```
Comment (joeauyeung) [inline packages/app-store/stripepayment/lib/constants.ts:14]: By having the payment options under each payment app package, we don't have to worry if an app doesn't support holds or future payment options.

File: packages/ui/components/form/checkbox/Checkbox.tsx [hunk 5]
```diff
@@ -5,7 +5,7 @@ import classNames from "@calcom/lib/classNames";
 
 type Props = InputHTMLAttributes<HTMLInputElement> & {
   label?: React.ReactNode;
-  description: string;
+  description: string | React.ReactNode;
```
Comment (joeauyeung) [inline packages/ui/components/form/checkbox/Checkbox.tsx:8]: I added the `React.ReactNode` type to `description` because I wanted to pass a `<Trans>` component. @sean-brydon

File: apps/web/components/booking/BookingDescriptionPayment.tsx
Comment (joeauyeung) [inline apps/web/components/booking/BookingDescriptionPayment.tsx:1]: This file component was used to format the price to the user's locale. I noticed throughout the code base that we use `IntlProvider` & `FormattedNumber`. i18n uses the javascript native format number w/ currency. 

Would replacing those two imports reduce our bundle size? If so, I would like to replace instances of `IntlProvider`

File: apps/web/components/booking/BookingListItem.tsx [hunk 6]
```diff
@@ -184,6 +196,18 @@ function BookingListItem(booking: BookingItemProps) {
...
+      id: "charge_card",
+      label: cardCharged ? t("no_show_fee_charged") : t("collect_no_show_fee"),
+      disabled: cardCharged,
+      onClick: () => {
+        setChargeCardDialogIsOpen(true);
+      },
+      icon: FiCreditCard,
+    },
+  ];
+
```
Comment (joeauyeung) [inline apps/web/components/booking/BookingListItem.tsx:210]: Adds the ability to collect the no show fee from Cal rather than going into Stripe

image

File: apps/web/components/dialog/ChargeCardDialog.tsx [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+                <FiAlertTriangle className="mr-2 h-5 w-5 " aria-hidden="true" />
+                <p className="text-sm">{t("error_charging_card")}</p>
+              </div>
+            )}
+
+            <DialogFooter>
+              <DialogClose />
+              <Button
+                data-testid="send_request"
+                disabled={chargeCardMutation.isLoading || chargeError}
```
Comment (joeauyeung) [inline apps/web/components/dialog/ChargeCardDialog.tsx:82]: If the charge fails, the user shouldn't be able to spam the button. By having it disabled until the dialog is closed or the refresh the page should help with that

File: apps/web/components/dialog/ChargeCardDialog.tsx
Comment (joeauyeung) [inline apps/web/components/dialog/ChargeCardDialog.tsx:1]: image

File: packages/app-store/stripepayment/lib/PaymentService.ts [hunk 8]
```diff
@@ -57,11 +68,67 @@ export class PaymentService implements IAbstractPaymentService {
       );
       const paymentFee = Math.round(payment.amount * payment_fee_percentage + payment_fee_fixed);
 
+      const customer = await retrieveOrCreateStripeCustomerByEmail(
+        bookerEmail,
+        this.credentials.stripe_user_id
+      );
+
```
Comment (joeauyeung) [inline packages/app-store/stripepayment/lib/PaymentService.ts:75]: Setup intents must be attached to a customer. Right now we do not attach any payments to a Stripe customer

File: packages/app-store/stripepayment/lib/PaymentService.ts [hunk 9]
```diff
@@ -92,11 +159,72 @@ export class PaymentService implements IAbstractPaymentService {
...
+        throw new Error();
+      }
+      return paymentData;
+    } catch (error) {
+      console.error(error);
+      throw new Error("Payment could not be created");
+    }
+  }
+
+  async chargeCard(payment: Payment, bookingId: Booking["id"]): Promise<Payment> {
```
Comment (joeauyeung) [inline packages/app-store/stripepayment/lib/PaymentService.ts:175]: This method should charge a stored payment method, so the user does not need to enter their payment method again

File: packages/app-store/stripepayment/lib/PaymentService.ts [hunk 10]
```diff
@@ -92,11 +159,72 @@ export class PaymentService implements IAbstractPaymentService {
...
+      const paymentFee = Math.round(payment.amount * payment_fee_percentage + payment_fee_fixed);
+
+      const params = {
+        amount: payment.amount,
+        currency: payment.currency,
+        application_fee_amount: paymentFee,
+        customer: setupIntent.customer,
+        payment_method: setupIntent.payment_method,
+        off_session: true,
+        confirm: true,
```
Comment (joeauyeung) [inline packages/app-store/stripepayment/lib/PaymentService.ts:202]: With these params, we can automatically complete the payment intent with the customer's stored payment method

File: packages/app-store/stripepayment/lib/PaymentService.ts [hunk 11]
```diff
@@ -92,11 +159,72 @@ export class PaymentService implements IAbstractPaymentService {
...
+
+      const paymentData = await prisma.payment.update({
+        where: {
+          id: payment.id,
+        },
+        data: {
+          success: true,
+          data: {
+            ...payment.data,
+            paymentIntent,
```
Comment (joeauyeung) [inline packages/app-store/stripepayment/lib/PaymentService.ts:219]: The stored data shouldn't replace the stored setup intent data we have in the DB

File: packages/app-store/stripepayment/lib/customer.ts [hunk 12]
```diff
@@ -82,3 +82,28 @@ export async function deleteStripeCustomer(user: UserType): Promise<string | nul
...
+  console.log("🚀 ~ file: customer.ts:87 ~ retrieveOrCreateStripeCustomerByEmail ~ email:", email);
+  const customer = await stripe.customers.list(
+    {
+      email,
+      limit: 1,
+    },
+    {
+      stripeAccount: stripeAccountId,
+    }
+  );
```
Comment (joeauyeung) [inline packages/app-store/stripepayment/lib/customer.ts:96]: This customer data isn't crucial like our team's customers as we were never collecting this data before. We need a customer to attach a payment method to

File: packages/emails/src/templates/AttendeeAwaitingPaymentEmail.tsx
Comment (joeauyeung) [inline packages/emails/src/templates/AttendeeAwaitingPaymentEmail.tsx:1]: image

File: packages/emails/src/templates/BaseScheduledEmail.tsx [hunk 13]
```diff
@@ -80,6 +80,16 @@ export const BaseScheduledEmail = (
...
+      {props.calEvent?.paymentInfo && (
+        <Info
+          label={props.calEvent.paymentInfo?.paymentOption === "HOLD" ? t("no_show_fee") : t("price")}
+          description={new Intl.NumberFormat(props.attendee.language.locale, {
+            style: "currency",
+            currency: props.calEvent.paymentInfo?.currency || "USD",
+          }).format(props.calEvent.paymentInfo?.amount / 100.0)}
+          withSpacer
+        />
+      )}
```
Comment (joeauyeung) [inline packages/emails/src/templates/BaseScheduledEmail.tsx:92]: When a booking has a price, we should also display it in the email

File: packages/emails/src/templates/NoShowFeeChargedEmail.tsx [hunk 14]
```diff
@@ -0,0 +1,33 @@
...
+  return (
+    <BaseScheduledEmail
+      title={t("no_show_fee_charged_text_body")}
+      headerType="calendarCircle"
+      subtitle={
+        <>
+          {t("no_show_fee_charged_subtitle", {
+            amount: calEvent.paymentInfo.amount / 100,
+            formatParams: { amount: { currency: calEvent.paymentInfo?.currency } },
+          })}
```
Comment (joeauyeung) [inline packages/emails/src/templates/NoShowFeeChargedEmail.tsx:25]: This is an example of how we can use i18n to format a number w/ currency.

File: packages/features/bookings/lib/handleNewBooking.ts [hunk 15]
```diff
@@ -1802,17 +1803,18 @@ async function handler(
     }
   }
 
-  if (!isConfirmedByDefault && noEmail !== true) {
+  const bookingRequiresPayment =
+    !Number.isNaN(paymentAppData.price) &&
+    paymentAppData.price > 0 &&
+    !originalRescheduledBooking?.paid &&
+    !!booking;
```
Comment (joeauyeung) [inline packages/features/bookings/lib/handleNewBooking.ts:1810]: DRY, since we're using this logic in more than one place we should abstract it

File: packages/features/bookings/lib/handleNewBooking.ts [hunk 16]
```diff
@@ -1802,17 +1803,18 @@ async function handler(
...
+
+  if (!isConfirmedByDefault && noEmail !== true && !bookingRequiresPayment) {
     await sendOrganizerRequestEmail({ ...evt, additionalNotes });
     await sendAttendeeRequestEmail({ ...evt, additionalNotes }, attendeesList[0]);
   }
```
Comment (joeauyeung) [inline packages/features/bookings/lib/handleNewBooking.ts:1815]: This was sending these emails before the payment was made. We now check if the booking requires payment before sending.

File: packages/features/ee/payments/api/webhook.ts [hunk 17]
```diff
@@ -34,6 +34,91 @@ async function getEventType(id: number) {
   });
 }
 
+async function getBooking(bookingId: number) {
```
Comment (joeauyeung) [inline packages/features/ee/payments/api/webhook.ts:37]: DRY again. Will reuse this logic for `payment_intent.succeeded` and `setup_intent.succeeded` events

File: packages/features/ee/payments/api/webhook.ts [hunk 18]
```diff
@@ -174,10 +259,56 @@ async function handlePaymentSuccess(event: Stripe.Event) {
...
+        update: {
+          ...bookingData,
+        },
+      },
+    },
+  });
+
+  // If the card information was already captured in the same customer. Delete the previous payment method
+
+  if (!eventTypeRaw?.requiresConfirmation) await sendScheduledEmails({ ...evt });
```
Comment (joeauyeung) [inline packages/features/ee/payments/api/webhook.ts:304]: As part of both events, should we delete all bookings in the same timeslot & event type that were awaiting payment?

Comment (joeauyeung) [inline packages/features/ee/payments/api/webhook.ts:304]: TODO: handle if the booking still requires confirmation

File: packages/features/ee/payments/components/Payment.tsx
Comment (joeauyeung) [inline packages/features/ee/payments/components/Payment.tsx:1]: image

File: packages/features/ee/payments/components/Payment.tsx [hunk 19]
```diff
@@ -70,22 +78,33 @@ export default function PaymentComponent(props: Props) {
...
     if (!card) return;
     setState({ status: "processing" });
-    const payload = await stripe.confirmCardPayment(props.payment.data.client_secret!, {
-      payment_method: {
-        card,
-      },
-    });
-    if (payload.error) {
+
+    let payload;
```
Comment (joeauyeung) [inline packages/features/ee/payments/components/Payment.tsx:82]: Why do we rely on webhooks for post purchase logic? Couldn't we pass the payload to a trpc endpoint?

File: packages/trpc/server/routers/viewer/payments.tsx [hunk 20]
```diff
@@ -0,0 +1,110 @@
...
+import { sendNoShowFeeChargedEmail } from "@calcom/emails";
+import { getTranslation } from "@calcom/lib/server/i18n";
+import type { CalendarEvent } from "@calcom/types/Calendar";
+
+import { TRPCError } from "@trpc/server";
+
+import { router, authedProcedure } from "../../trpc";
+
+export const paymentsRouter = router({
+  chargeCard: authedProcedure
```
Comment (joeauyeung) [inline packages/trpc/server/routers/viewer/payments.tsx:14]: End point to charge the no show fee

File: packages/trpc/server/routers/viewer/payments.tsx [hunk 21]
```diff
@@ -0,0 +1,110 @@
...
+          appId: booking.payment[0].appId,
+        },
+        include: {
+          app: true,
+        },
+      });
+
+      const paymentApp = appStore[paymentCredential?.app?.dirName as keyof typeof appStore];
+      const PaymentService = paymentApp.lib.PaymentService;
+      const paymentInstance = new PaymentService(paymentCredential);
```
Comment (joeauyeung) [inline packages/trpc/server/routers/viewer/payments.tsx:91]: @alannnc I love the payment service you created and I'm excited to see new payment apps

**Comment by keithwillcode** [comment]
@joeauyeung Is there a mechanism built here that prevents multiple collections of the no-show fee? I saw the Loom you made where the button on the UI gets disabled after one is collected but just wanted to make sure we have server-side logic that prevents multiple charges. Didn't see one but here might have just missed it. 🙏

File: apps/web/components/booking/BookingDescriptionPayment.tsx [hunk 22]
```diff
@@ -1,22 +1,35 @@
...
+        </>
+      ) : (
+        <IntlProvider locale="en">
+          <FormattedNumber
+            value={paymentAppData.price / 100.0}
+            style="currency"
+            currency={paymentAppData.currency?.toUpperCase()}
+          />
+        </IntlProvider>
+      )}
```
Comment (joeauyeung) [inline apps/web/components/booking/BookingDescriptionPayment.tsx:32]: For "HOLD" I used i18n to format the price instead of the `IntlProvider`. Could we remove `intlProvider` through out the codebase and use i18n instead to reduce our bundle size?

File: packages/app-store/google-tag-manager/config.json [hunk 23]
```diff
@@ -17,12 +15,12 @@
...
         }
       ]
     }
   },
   "isTemplate": false,
   "__createdUsingCli": true,
   "__template": "booking-pages-tag"
-}
\ No newline at end of file
+}
```
Comment (alannnc) [inline packages/app-store/google-tag-manager/config.json:26]: We should undo this changes

File: packages/prisma/schema.prisma [hunk 24]
```diff
@@ -412,19 +412,25 @@ model ReminderMail {
...
+  booking       Booking?       @relation(fields: [bookingId], references: [id], onDelete: Cascade)
+  amount        Int
+  fee           Int
+  currency      String
+  success       Boolean
+  refunded      Boolean
+  data          Json
+  externalId    String         @unique
+  paymentOption PaymentOption? @default(ON_BOOKING)
+}
```
Comment (alannnc) [inline packages/prisma/schema.prisma:429]: Some extra indenting here. ^

File: packages/prisma/seed-app-store.config.json [hunk 25]
```diff
@@ -2,300 +2,228 @@
...
-      "analytics"
-    ],
+    "categories": ["analytics"],
     "slug": "google-tag-manager",
     "type": "google-tag-manager_analytics",
     "isTemplate": false
   }
-]
\ No newline at end of file
+]
```
Comment (alannnc) [inline packages/prisma/seed-app-store.config.json:229]: undo

File: packages/trpc/server/routers/viewer/bookings.tsx [hunk 26]
```diff
@@ -268,6 +268,14 @@ export const bookingsRouter = router({
           },
           status: true,
           paid: true,
+          payment: {
```
Comment (emrysal) [inline packages/trpc/server/routers/viewer/bookings.tsx:271]: This Query is too big nowadays - needs refactoring (potentially design wise, even) to optimise this query.

File: apps/web/components/dialog/ChargeCardDialog.tsx [hunk 27]
```diff
@@ -0,0 +1,100 @@
...
+  return (
+    <Dialog open={isOpenDialog} onOpenChange={setIsOpenDialog}>
+      <DialogContent>
+        <div className="flex flex-row space-x-3">
+          <div className="flex h-10 w-10 flex-shrink-0 justify-center rounded-full bg-[#FAFAFA]">
+            <FiCreditCard className="m-auto h-6 w-6" />
+          </div>
+          <div className="pt-1">
+            <DialogHeader title={t("charge_card")} />
+            <Trans i18nKey="charge_card_dialog_body">
```
Comment (emrysal) [inline apps/web/components/dialog/ChargeCardDialog.tsx:52]: I think this translation is missing

File: apps/web/components/dialog/ChargeCardDialog.tsx [hunk 28]
```diff
@@ -0,0 +1,100 @@
...
+              <p className="text-sm text-gray-500">
+                You are about to charge the user{" "}
```
Comment (emrysal) [inline apps/web/components/dialog/ChargeCardDialog.tsx:54]: You are about to charge the attendee
```suggestion
                You are about to charge the attendee{" "}
```

**Review by emrysal — COMMENTED** [review]
Some i18n suggestions

File: apps/web/components/booking/BookingDescriptionPayment.tsx (hunk 22 above)
Comment (emrysal) [reply apps/web/components/booking/BookingDescriptionPayment.tsx:32]: This is fine, also hardcoded to "en" is a bug I believe elsewhere, instead i18n.language should be used.

File: packages/trpc/server/routers/viewer/bookings.tsx (hunk 26 above)
Comment (emrysal) [reply packages/trpc/server/routers/viewer/bookings.tsx:271]: As discussed, ignore for now.

File: apps/web/components/dialog/ChargeCardDialog.tsx
Comment (emrysal) [reply apps/web/components/dialog/ChargeCardDialog.tsx:1]: I think this error should be a toast, this dialog should be closed on failure as there's nothing you can do in the dialog to fix.

File: packages/features/ee/payments/api/webhook.ts [hunk 29]
```diff
@@ -180,10 +265,56 @@ async function handlePaymentSuccess(event: Stripe.Event) {
...
+      externalId: setupIntent.id,
+    },
+  });
+
+  if (!payment?.data || !payment?.id) throw new HttpCode({ statusCode: 204, message: "Payment not found" });
+
+  const { booking, user, evt, eventTypeRaw } = await getBooking(payment.bookingId);
+
+  const bookingData: Prisma.BookingUpdateInput = {
+    paid: true,
```
Comment (emrysal) [inline packages/features/ee/payments/api/webhook.ts:281]: This may be incorrect 💭

File: packages/features/ee/payments/components/Payment.tsx (hunk 19 above)
Comment (emrysal) [reply packages/features/ee/payments/components/Payment.tsx:79]: Not needed, in this situation it only complicates matters.

**Review by emrysal — CHANGES_REQUESTED** [review]
Great work! Left some comments to address & the seats function needs to be actively prevented from using this function until we can.

File: apps/web/components/dialog/ChargeCardDialog.tsx (hunk 27 above)
Comment (joeauyeung) [reply apps/web/components/dialog/ChargeCardDialog.tsx:52]: With the `<Trans>` component, I thought we didn't have to add the translation to the `common.json` file. For example, `apps/web/pages/settings/my-account/calendars.tsx` line 98.

File: packages/app-store/stripepayment/lib/PaymentService.ts [hunk 30]
```diff
@@ -104,6 +167,80 @@ export class PaymentService implements IAbstractPaymentService {
...
+          success: true,
+          data: {
+            ...paymentObject,
+            paymentIntent,
+          } as unknown as Prisma.InputJsonValue,
+        },
+      });
+
+      if (!paymentData) {
+        throw new Error();
```
[the remark this answers is not in the document]
Comment (joeauyeung) [reply packages/app-store/stripepayment/lib/PaymentService.ts:232]: If this does happen that means the paymentIntent failed so we don't need to revert but if we tried again a new paymentIntent would be created.

File: packages/app-store/stripepayment/lib/PaymentService.ts [hunk 31]
```diff
@@ -38,9 +40,14 @@ export class PaymentService implements IAbstractPaymentService {
...
   async create(
     payment: Pick<Prisma.PaymentUncheckedCreateInput, "amount" | "currency">,
-    bookingId: Booking["id"]
+    bookingId: Booking["id"],
+    bookerEmail: string,
+    paymentOption: PaymentOption
   ) {
     try {
+      // Ensure that the payment service can support the passed payment option
+      paymentOptionEnum.parse(paymentOption);
```
[the remark this answers is not in the document]
Comment (joeauyeung) [reply packages/app-store/stripepayment/lib/PaymentService.ts]: I created a charge on booking and collecting the payment method as it's own method under the PaymentService class

https://github.com/calcom/cal.com/pull/8022/commits/ec742df46b7aa0755191def210cbc1ac70feb1fd

File: packages/features/ee/payments/api/webhook.ts (hunk 29 above)
Comment (joeauyeung) [reply packages/features/ee/payments/api/webhook.ts:281]: On line 317, this function only triggers when we get a setup intent succeed event from Stripe

**Comment by joeauyeung** [comment]
https://github.com/calcom/cal.com/pull/8022#discussion_r1158650184

In this case, we're calling two different methods `confirmCardSetup` and `confirmCardPayment`. Would a try / catch block make things simpler?

**Comment by emrysal** [comment]
image

This error shows up for "Collect payment on booking" too with seats enabled. It should only show when selecting "Collect on no-show"

**Comment by joeauyeung** [comment]
> As the organizer, I'm stuck @ image

@emrysal are you forwarding Stripe webhooks to your local host? With the Stripe CLI you can use `stripe listen --forward-to localhost:3000/api/integrations/stripepayment/webhook`

**Review by emrysal — APPROVED** [review]
